src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/GeneratorDiagnostic.cs

```diff
@@ -55,8 +57,130 @@ public override DiagnosticInfo ToDiagnosticInfo(DiagnosticDescriptor descriptor,
                     UnnecessaryDataLocations,
                     // Add "unnecessary locations" property so the IDE fades the right locations.
                     DiagnosticProperties.Add(WellKnownDiagnosticTags.Unnecessary, $"[{string.Join(",", Enumerable.Range(0, UnnecessaryDataLocations.Length))}]"),
-                    UnnecessaryDataDetails,
-                    elementName);
+                    UnnecessaryDataName,
+                    elementName,
+                    UnnecessaryDataDetails ?? "");
+            }
+        }
+
+        /// <summary>
+        /// Provides an implementation of <see cref="IMarshallingGenerator.SupportsByValueMarshalKind(ByValueContentsMarshalKind, TypePositionInfo, StubCodeContext, out GeneratorDiagnostic?)"/> through <see cref="GetSupport(ByValueContentsMarshalKind, TypePositionInfo, StubCodeContext, out GeneratorDiagnostic?)"/>
+        /// </summary>
+        public record ByValueMarshalKindSupportManager(
+            ByValueMarshalKindSupport InSupport, string? InSupportDetails,
+            ByValueMarshalKindSupport OutSupport, string? OutSupportDetails,
+            ByValueMarshalKindSupport InOutSupport, string? InOutSupportDetails)
+        {
+            /// <summary>
+            /// A default <see cref="ByValueMarshalKindSupportManager"/> for by value parameters. [In] is allowed, but unnecessary. Out is not allowed.
+            /// </summary>
+            public static ByValueMarshalKindSupportManager ValueTypeParameterDefault = new ByValueMarshalKindSupportManager(
+                InSupport: ByValueMarshalKindSupport.Unnecessary, InSupportDetails: SR.InAttributeOnlyIsDefault,
+                OutSupport: ByValueMarshalKindSupport.NotSupported, OutSupportDetails: SR.OutAttributeNotSupportedOnByValueParameters,
+                InOutSupport: ByValueMarshalKindSupport.NotSupported, InOutSupportDetails: SR.OutAttributeNotSupportedOnByValueParameters);
+
+            /// <summary>
+            /// A default <see cref="ByValueMarshalKindSupportManager"/> for by reference parameters. [In] is allowed, but unnecessary. Out is allowed.
+            /// </summary>
+            public static ByValueMarshalKindSupportManager ReferenceTypeParameterDefault = new ByValueMarshalKindSupportManager(
```

Comment:
Is this for reference type parameters or for parameters passed by reference? If it's for "by reference" parameters, can we rename this to `ByReferenceParameterDefault`?

If not, I think these defaults need to be revisited. It's not whether a parameter type is a struct or a class that determines if `[Out]` is valid or not, but whether or not the parameter type represents a collection type (in particular, we only want to allow it for arrays initially as we try to limit usages of these attributes).

Reply:
This is for reference type parameters. Why do we want to limit the use of [Out] on classes? Doesn't it just mean the callee may modify a value and we need to marshal it back?

Reply:
In regular managed code, that's true, but in interop, we have more restrictive rules.

In the built-in interop system, that is true for "layout classes" (classes with `LayoutKind.Sequential` or `LayoutKind.Explicit`), but we don't support that marshalling model in the new interop system as it was a very confusing feature that caused a lot of pain. The only case where passing an object by value that is expected to be modifiable by the callee is collections (where the elements can be modified, but not the collection itself).

For example, you can't change the function pointer for a delegate type or the handle value of a SafeHandle when passing them by value.